### ISSUE TITLE
Selected cell position is wrong on in/out dates

### ISSUE THREAD

**Issue by danlozano** [patchthecode/JTAppleCalendar#391]
I'm using the newest version 7.0.

I'm using the CellState's selected position to draw the cells. It works almost perfectly on the new version. But I realized that the behavior is kind of unexpected when the calendar has in/out dates.

For example:
<img width="200" alt="screenshot1" src="https://cloud.githubusercontent.com/assets/2406956/25398467/dc6c11c8-29b1-11e7-823f-4b609d7bcc72.png">

Here the selected range is from the 25th to the 6th of the next month. Since I'm using in and out dates the row from the 30th to the 6th are repeated. I have no issue with this. BUT the selectedPosition is unexpected. If the 6th is the RIGHT selectedPosition I would expect it to be for BOTH rows. Because if I'm looking just at the previous month, without going to the next one, it will look wrong.

Also these other examples:

<img width="200" alt="screenshot2" src="https://cloud.githubusercontent.com/assets/2406956/25398552/2165c634-29b2-11e7-88d8-6d41c910be88.png">

Here the 30th should be a LEFT selectedPosition.

--- 

<img width="200" alt="screenshot3" src="https://cloud.githubusercontent.com/assets/2406956/25398582/363d0874-29b2-11e7-9578-fd4283caaed3.png">
Here the 6th should be a MIDDLE but since it's at the end of the month it's considered a RIGHT which in my opinion is wrong.

--- 

Any ideas about this? I think in the previous versions I did not have this problem.

Thanks for the great framework!

**Comment by patchthecode** [comment]
ok will check this out after work

**Comment by patchthecode** [comment]
Alright the logic that you see, is what I have coded it to be.
Selected position tells you exactly, the _selectedPosition_

In the 1st image the `6th` outDate is between the the 5th selected cell and the 30th selected cell. Therefore it is middle.
<img width="200" alt="screenshot1" src="https://cloud.githubusercontent.com/assets/2406956/25398467/dc6c11c8-29b1-11e7-823f-4b609d7bcc72.png">

in the 2nd image the `30th` selected cell, is between the 6th selected cell and the 1st selected cell, therefore it's selected position is middle.
<img width="200" alt="screenshot2" src="https://cloud.githubusercontent.com/assets/2406956/25398552/2165c634-29b2-11e7-88d8-6d41c910be88.png">

in the 3rd image, The `6th` selected cell has one cell to its left that is selected, but there is no cell to its right that is selected. Therefore it's selected position is right.
<img width="200" alt="screenshot3" src="https://cloud.githubusercontent.com/assets/2406956/25398582/363d0874-29b2-11e7-9578-fd4283caaed3.png">

SelectedPosition literally describes its _selectedPosition_.

But I understand the problem you are having with this and I think I can change the logic to reflect what you have explained. I am only concerned if this behavior change will affect other people negatively.

**Comment by danlozano** [comment]
Hi. Thanks for the quick response!

I get what you are saying. If the selectedPosition is just literally comparing index paths, without any semantics, then yes they are correct.

But the ranges are date ranges, and if you look at them that way the selectedPosition's are wrong when using in/out dates. 

If a user is only looking at one month, MAY, and selects a range from any day to the last visible day in that month (which would be an out date) the end will appear as middle, it will not appear as a range, because the actual RIGHT selectedPosition is in the other month's in dates. That makes no sense.

I think I did not have this problem in the previous versions that's why I created the issue. But I guess I'll fix them in my own cell display logic but they seem wrong to me.

Thanks for the great library!

**Comment by patchthecode** [comment]
Well like i said, what you have explained makes sense.
I am considering to fix this to the way that you have described, because it makes sense.

**Comment by danlozano** [comment]
Yeah, I guess it's a typical problem in computer science. Both ways are "correct" depending on the way you look at it hehe.

I just want what's best for the framework in general, and I think in this case basing this on dates and not on exact index paths would make sense. 

I also have to point out that I think this problem only pops up when using in/out dates.

For now what I'm doing is "fixing" the positions myself doing date comparisons.

```swift
switch cellState.selectedPosition() {
        case .full:
            setFull(selectedView: myCustomCell.selectedView)
        case .left:
            if isYesterdaySelected(cellState.date) {
                setMiddle(selectedView: myCustomCell.selectedView)
            } else {
                setLeft(selectedView: myCustomCell.selectedView)
            }
        case .right:
            if isTomorrowSelected(cellState.date) {
                setMiddle(selectedView: myCustomCell.selectedView)
            } else {
                setRight(selectedView: myCustomCell.selectedView)
            }
        case .middle:
            if isDateLastDateSelected(cellState.date) {
                setRight(selectedView: myCustomCell.selectedView)
            } else if isDateFirstDateSelected(cellState.date) {
                setLeft(selectedView: myCustomCell.selectedView)
            } else {
                setMiddle(selectedView: myCustomCell.selectedView)
            }
        case .none:
            setNone(selectedView: myCustomCell.selectedView)
        }
```

That's awesome @patchthecode ! Thanks for the great library and quick response. I'll wait and see what you decide & come up with. If you need any help let me know.

**Comment by danlozano** [comment]
My only fear with my solution is that it might affect performance. 🤔

**Comment by danlozano** [comment]
Maybe this feature could be enabled/disabled via a property.

```swift
public var treatSectionBoundariesAsDateBoundaries = true // Default = true, would be how it works today.
```

Because, in my case, since I have in/out dates, and I move a month at a time, it makes total sense to have it work how I want it to.

But maybe if someone has smooth scrolling without stopping at each month this might look weird. 
![simulator screen shot apr 26 2017 12 03 57 pm](https://cloud.githubusercontent.com/assets/2406956/25446839/7219a75a-2a78-11e7-9072-d268168c9244.png)

🤔

**Comment by patchthecode** [comment]
Haha. Let's see. All take a deep look at it over the weekend.
I'll also take a look at the sample code you posted above.

**Comment by patchthecode** [comment]
this has now been fixed on master branch

**Comment by patchthecode** [comment]
closing issue.
will be included in release 7.1.0
